components/brave_wallet/browser/keyring_controller.cc

```diff
@@ -675,6 +720,107 @@ std::vector<mojom::AccountInfoPtr> KeyringController::GetAccountInfosForKeyring(
   return result;
 }
 
+void KeyringController::GetHardwareAccounts(
+    GetHardwareAccountsCallback callback) {
+  std::vector<mojom::HardwareWalletAccountPtr> accounts;
+  base::Value hardware_keyrings(base::Value::Type::DICTIONARY);
+  const base::Value* value = GetPrefForHardwareKeyring(prefs_);
+  if (!value) {
+    std::move(callback).Run({});
+    return;
+  }
+
+  for (const auto hw_keyring : value->DictItems()) {
+    std::string device_id = hw_keyring.first;
+    const base::Value* account_value =
+        hw_keyring.second.FindKey(kAccountMetas);
+    if (!account_value)
+      continue;
+    SerializeHardwareAccounts(account_value, &accounts);
+  }
+
+  std::move(callback).Run(std::move(accounts));
+}
+
+// static
+const base::Value* KeyringController::GetPrefForHardwareKeyring(
+    PrefService* prefs) {
+  DCHECK(prefs);
+  const base::DictionaryValue* keyrings_pref =
+      prefs->GetDictionary(kBraveWalletKeyrings);
+  if (!keyrings_pref)
+    return nullptr;
+  return keyrings_pref->FindKey(kHardwareKeyrings);
+}
+
+void KeyringController::AddHardwareAccounts(
+    std::vector<mojom::HardwareWalletAccountPtr> infos) {
+  if (infos.empty())
+    return;
+  const auto& hardware = infos.front()->hardware;
+  const auto hash = base::PersistentHash(infos.front()->address);
+  std::string device_id = hardware + std::to_string(hash);
+
+  base::Value hardware_keyrings(base::Value::Type::DICTIONARY);
+  const base::Value* value = GetPrefForHardwareKeyring(prefs_);
+  if (value) {
+    hardware_keyrings = value->Clone();
+  }
+  const base::Value* device_value = hardware_keyrings.FindKey(device_id);
+  base::Value device_keyring(base::Value::Type::DICTIONARY);
+  if (device_value)
+    device_keyring = device_value->Clone();
+
+  const base::Value* meta_value = device_keyring.FindKey(kAccountMetas);
+  base::Value account_meta(base::Value::Type::DICTIONARY);
+  if (meta_value)
+    account_meta = meta_value->Clone();
+
+  for (const auto& info : infos) {
+    DCHECK_EQ(hardware, info->hardware);
+    if (hardware != info->hardware)
+      continue;
+    base::Value hw_account(base::Value::Type::DICTIONARY);
+    hw_account.SetStringKey(kAccountName, info->name);
+    hw_account.SetStringKey(kAccountHardware, info->hardware);
+    hw_account.SetStringKey(kHardwareDerivationPath, info->derivation_path);
+
+    account_meta.SetKey(info->address, std::move(hw_account));
+  }
+
+  device_keyring.SetKey(kAccountMetas, std::move(account_meta));
+  hardware_keyrings.SetKey(device_id, std::move(device_keyring));
+
+  DictionaryPrefUpdate update(prefs_, kBraveWalletKeyrings);
+  base::DictionaryValue* keyrings_pref = update.Get();
+  keyrings_pref->SetKey(kHardwareKeyrings, std::move(hardware_keyrings));
```

Comment:
we need to call `NotifyAccountsChanged()` so UI can fetch new info when a new hardware account is added

Reply:
done